src/Utilities/RunTimeManager.cpp

```diff
@@ -26,39 +27,127 @@ template class RunTimeManager<CPUClock>;
 {
   if (nloop > 0)
-  {
     return total_time / nloop;
-  }
   return 0.0;
 }
 
 template class LoopTimer<CPUClock>;
 template class LoopTimer<FakeCPUClock>;
 
+template<class CLOCK>
+RunTimeControl<CLOCK>::RunTimeControl(RunTimeManager<CLOCK>& rm,
+                                      int maxCPUSecs,
+                                      const std::string& stop_file_prefix,
+                                      bool cleanup)
+    : MaxCPUSecs(maxCPUSecs),
+      runtimeManager(rm),
+      stop_filename_(stop_file_prefix + ".STOP"),
+      stop_status_(StopStatus::CONTINUE)
+{
+  if (stop_file_prefix.empty())
+    throw std::runtime_error("Stop control filename prefix must not be empty!");
+
+  if (cleanup)
+  {
+    std::remove(stop_filename_.c_str());
+    if (std::ifstream(stop_filename_.c_str()))
+      throw std::runtime_error("Failed to delete the existing stop control file \"" + stop_filename_ +
+                               "\", cannot continue!");
+  }
+
+  m_runtime_safety_padding = 10.0; // 10 seconds - enough to shut down?
+  m_loop_margin            = 1.1;  // 10% margin on average loop time?
+}
+
 template<class CLOCK>
 bool RunTimeControl<CLOCK>::enough_time_for_next_iteration(LoopTimer<CLOCK>& loop_timer)
 {
-  m_loop_time      = loop_timer.get_time_per_iteration();
-  m_elapsed        = runtimeManager.elapsed();
+  m_loop_time = loop_timer.get_time_per_iteration();
+  m_elapsed   = runtimeManager.elapsed();
+
+  if (m_elapsed >= MaxCPUSecs)
+  {
+    stop_status_ = StopStatus::MAX_SECONDS_PASSED;
+    return false;
+  }
+
   m_remaining      = MaxCPUSecs - m_elapsed;
   bool enough_time = true;
   if ((m_loop_margin * m_loop_time + m_runtime_safety_padding) > m_remaining)
     enough_time = false;
 
+  stop_status_ = StopStatus::NOT_ENOUGH_TIME;
   return enough_time;
 }
 
 template<class CLOCK>
-std::string RunTimeControl<CLOCK>::time_limit_message(const std::string& driverName, int block)
+bool RunTimeControl<CLOCK>::stop_file_reqeusted()
+{
+  if (std::ifstream(stop_filename_.c_str()))
+  {
+    stop_status_ = StopStatus::STOP_FILE;
+    return true;
+  }
+  else
+    return false;
+}
+
+template<class CLOCK>
+bool RunTimeControl<CLOCK>::checkStop(LoopTimer<CLOCK>& loop_timer)
+{
+  bool need_to_stop = false;
+  need_to_stop |= !enough_time_for_next_iteration(loop_timer);
+  need_to_stop |= stop_file_reqeusted();
+  return need_to_stop;
+}
+
+template<class CLOCK>
+std::string RunTimeControl<CLOCK>::generateStopMessage(const std::string& driverName, int block) const
 {
   std::stringstream log;
-  log << "Time limit reached for " << driverName << ", stopping after block " << block - 1 << std::endl;
-  log << "  Iteration time per " << driverName << " block (seconds) = " << m_loop_time << std::endl;
-  log << "  Elapsed time (seconds)       = " << m_elapsed << std::endl;
-  log << "  Remaining time (seconds)      = " << m_remaining << std::endl;
+  log << "RunTimeControl in " << driverName << " kicks in." << std::endl;
```

Comment:
Message should be more formal than "kicks in".
Maybe "RunTimeControl activated in <drivername>"

Reply:
Changed to `"RunTimeControl takes action in " << driverName << " driver."`